transactron/utils/amaranth_ext/memory.py

```diff
@@ -0,0 +1,368 @@
+            raise ValueError("Invalid port domain: Only synchronous memory ports supported.")
+        return ReadPort(
+            memory=self,
+            width=self.shape,
+            depth=self.depth,
+            init=self.init,
+            transparent_for=transparent_for,
+            src_loc=1 + src_loc_at,
+        )
+
+    def write_port(self, *, domain: str = "sync", granularity: Optional[int] = None, src_loc_at: int = 0):
+        if self._frozen:
+            raise AlreadyElaborated("Cannot add a memory port to a memory that has already been elaborated")
+        if domain != "sync":
+            raise ValueError("Invalid port domain: Only synchronous memory ports supported.")
+        return WritePort(
+            memory=self,
+            width=self.shape,
+            depth=self.depth,
+            init=self.init,
+            granularity=granularity,
+            src_loc=1 + src_loc_at,
+        )
+
+
+class MultiReadMemory(BaseMultiportMemory):
+    """Memory with one write and multiple read ports.
+
+    One can request multiple read ports and not more than 1 write port. Module internally
+    uses multiple (number of read ports) instances of amaranth.lib.memory.Memory with one
+    read and one write port.
+
+    """
+
+    def write_port(self, *, domain: str = "sync", granularity: Optional[int] = None, src_loc_at: int = 0):
+        if self.write_ports:
+            raise MultipleWritePorts("Cannot add multiple write ports to a single write memory")
+        return super().write_port(domain=domain, granularity=granularity, src_loc_at=src_loc_at)
+
+    def elaborate(self, platform):
+        m = Module()
+
+        self._frozen = True
+
+        write_port = self.write_ports[0] if self.write_ports else None
+        for port in self.read_ports:
+            # for each read port a new single port memory block is generated
+            mem = memory.Memory(
+                shape=port.width, depth=self.depth, init=self.init, attrs=self.attrs, src_loc_at=self.src_loc
+            )
+            m.submodules += mem
+            physical_write_port = mem.write_port(granularity=write_port.granularity) if write_port else None
+            physical_read_port = mem.read_port(
+                transparent_for=(
+                    [physical_write_port] if physical_write_port and write_port in port.transparent_for else []
+                )
+            )
+            m.d.comb += [
+                physical_read_port.addr.eq(port.addr),
+                port.data.eq(physical_read_port.data),
+                physical_read_port.en.eq(port.en),
+            ]
+
+            if physical_write_port and write_port:
+                m.d.comb += [
+                    physical_write_port.addr.eq(write_port.addr),
+                    physical_write_port.data.eq(write_port.data),
+                    physical_write_port.en.eq(write_port.en),
+                ]
+
+        return m
+
+
+class MultiportXORMemory(BaseMultiportMemory):
```

Comment:
Granularity not supported, it would be nice if this was documented and an exception was thrown if `granularity` is specified for any write port.